Import Missions with Occurrence Exceptions

achrafmohye opened the pull request:
Closes [#12341](https://github.com/OpenRailAssociation/osrd/issues/12341) [#12342](https://github.com/OpenRailAssociation/osrd/issues/12342)

achrafmohye commented:
I can give a file for testing with the exact cases depicted in the the screenshot of the original feature

RomainValls commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +125,197 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+      }
+
+      exceptions.push(exception);
+    }
```
this part breaks the app when importing some files, as we're always pushing exceptions even when the trainSchedule is identical to the model. In `useOccurences`, the exception expects a `start_time!.value` and here we're not providing it. 
I tried by preventing the exception to be pushed and it seems to work : 
```suggestion

if (!isIdenticalToModel) {
  const exception: PacedTrainException = {
    key: uuidV4(),
  };

  applyScheduleDifferencesToException(exception, viriatoSchedule, differences);
  exceptions.push(exception);
}

```

another solution could be to provide the exception with the start_time of the trainSchedule ?

achrafmohye replied:
Done :)

RomainValls commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
When trying to import other files than the example one, i have an error when starting the import : 
[0].exceptions[11].path_and_schedule.schedule[23].arrival: Failed to parse duration: -PT21H42M54S at line 1 column 58072"

achrafmohye replied:
Done :)

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -10,6 +17,52 @@ const extractCiChCode = (code: string) => {
+ * - `isIdenticalToModel`: A boolean indicating if the schedules are identical based on the defined criteria.
+ * - `differences`: A Set of `ScheduleDifferenceKey` enum values indicating which aspects of the schedules differ.
+ */
+export const getScheduleDifferences = (
```
@SharglutDev has already written a function which compares two train schedules and computes differences. Could we re-use the same functions here?

@SharglutDev, can you remind me the name of your helper functions?

SharglutDev replied:
`generatePacedTrainException` maybe ?

SharglutDev replied:
It compares more stuff than what @achrafmohye is, but maybe it can be adapt ?

emersion replied:
Right. That function computes differences between two `PacedTrain` entries. It doesn't access the `paced` field though, so could operate on `TrainSchedule` instead.

At the top of `updatePacedTrainExceptionsList()`, `hasExceptions` checks whether the exception is empty.

emersion replied:
> It compares more stuff than what @achrafmohye is, but maybe it can be adapt ?

I think we only compare stuff that we populate in the import here, so more stuff isn't harmful here.

maelysLeratRosso replied:
And it will make it painfree when we add more stuff to import (categories,...) !

achrafmohye replied:
Done :)

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/buildStepsFromOcp.ts:
```diff
@@ -100,7 +100,7 @@ export const buildSteps = (
       schedule.push({
         at: id,
-        arrival: Duration.subtractDate(step.arrivalDate, departureTime).toISOString(),
+        arrival: Duration.subtractDate(step.arrivalDate, departureTime).abs().toISOString(),
```
Hm, is `abs()` really something we should be doing here? It seems weird to me that the arrival time is before the departure time. This smells like a bug in the code above rather than here.

It sounds like this bug is orthogonal to your PR though, and that it just so happens that the particular file we're feeding triggers the bug.

achrafmohye replied:
I added a fix for this :)

emersion commented on front/src/utils/trainId.ts:
```diff
@@ -213,7 +213,7 @@ export const extractOccurrenceIndexFromOccurrenceId = (occurrenceId: OccurrenceI
   }
 
-  const formattedOccurrenceIndex = Number(occurrenceId.split('_')[2]);
+  const formattedOccurrenceIndex = Number(occurrenceId.split('_').at(-1));
```
What is the purpose of this change?

achrafmohye replied:
Sometimes the occurence id coming from the import can contain underscores so it's better, given how the index is contructed it's safer to take the last part as the index

emersion replied:
If we stop abusing occurrence IDs as discussed in https://github.com/OpenRailAssociation/osrd/pull/12498#discussion_r2240202205, this shouldn't be needed anymore.

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -1,15 +1,54 @@
   return { ciCode: Number(ciCode), chCode };
 };
 
+const createIndexedOccurrenceId = (pacedTrainId: string, index: number): IndexedOccurrenceId =>
+  `indexedoccurrence_${pacedTrainId}_${index}` as IndexedOccurrenceId;
+
+const createAddedExceptionId = (pacedTrainId: string): AddedExceptionId =>
+  `exception_${pacedTrainId}_${uuidV4()}` as AddedExceptionId;
```
We already have functions to create occurrence/exception IDs in `utils/trainId`. Can we avoid re-defining them here?

achrafmohye replied:
Here the "imported" `pacedTrainId` is not necessarily a number so we need to handled it differently, do you see another alternative ?
- I can move these helper functions `createIndexedOccurrenceId `/ `createAddedExceptionId` into `trainId.ts`

emersion replied:
Ah, sounds like we're abusing train IDs here. `IndexedOccurrenceId` and `AddedExceptionId` are never expected to contain a non-number paced train ID.

I'd suggest reworking `generatePacedTrainException()` so that we don't have to provide an ID. For instance, we can make `generatePacedTrainException()` return an `Omit<PacedTrainException, 'key' | 'occurrence_index'>` and leave it to the caller to generate these fields. The function can take an `occurrenceIndex: number | null` instead of an `OccurrenceId`.

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -1,15 +1,54 @@
+    'rolling_stock',
+    'start_time',
+    // add any other fields you want to check for an exception
+  ];
```
Do we really want to introduce a whitelist of fields to check?

Wouldn't it make more sense to have a blacklist of fields to ignore? That way we're less likely to forget about updating this spot once a new field is added to timetable items.

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -1,15 +1,54 @@
+
+const hasExceptions = (exception: PacedTrainException): boolean => {
+  // Remove train_name as we don't handle it as an exception
+  exception.train_name = undefined;
```
I don't think we need this since we got a whitelist below.

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +111,176 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+  );
+
+  sortedViriatoSchedules.forEach((viriatoSchedule) => {
+    if (unmatchedViriatoScheduleKeys.has(getScheduleKey(viriatoSchedule))) {
```
Style nit: an early return can be used to avoid increasing indentation too much.

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -1,15 +1,54 @@
+    // add any other fields you want to check for an exception
+  ];
+
+  return fieldsToCheck.some((field) => exception[field as keyof PacedTrainException]);
```
If we define `fieldsToCheck`'s type properly, we don't need this cast:

```ts
const fieldsToCheck: keyof PacedTrainException = [
```

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -138,11 +347,11 @@ const parseXML = async (xmlDoc: Document): Promise<TimetableJsonPayload> => {
       train_name: trainNumber,
       rolling_stock_name: rollingStockName || formationRef || '', // RollingStocks in xml files rarely have the correct format
       start_time: new Date(`${startDate} ${firstDepartureTimeformatted}`).toISOString(),
-      constraint_distribution: xmlDoc.documentElement.hasAttribute('xmlns:tps')
-        ? 'STANDARD'
-        : 'MARECO',
```
Why remove this?

achrafmohye replied:
Choice discussed with @Caracol3 and @maelysLeratRosso we want the default to be `MARECO` rather than `STANDARD`

maelysLeratRosso replied:
Oops I think there was a misunderstanding, we want STANDARD for both

emersion replied:
I think @achrafmohye made a typo in his comment: the code switches this to `STANDARD` unconditionally.

Cool, if that's something discussed with POs, that's completely fine by me - even allows us to remove a special case! (Ideally should be on its own commit since it's orthogonal to exceptions.)

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -138,11 +347,11 @@ const parseXML = async (xmlDoc: Document): Promise<TimetableJsonPayload> => {
+      constraint_distribution: 'STANDARD',
       path,
       schedule,
+      margins: { boundaries: [], values: ['0%'] },
+      power_restrictions: [],
```
Why do we need to add these fields?

achrafmohye replied:
These are required fields for the api call but we don't have these values in the type of the imported file

emersion replied:
Hm, I'm confused. The import code was working fine before this PR without these two fields. What has changed to make it so the API rejects payloads without these fields now?

achrafmohye replied:
Now we add exceptions to the call which was not handled before

emersion replied:
I'm still confused, sorry: power restrictions are not required by the API, be it for paced trains or for exceptions. Same goes for margins.

achrafmohye replied:
You're right that power restrictions are not required by the API, and this works as expected. However, even though margins are marked as optional, the API call fails if margins are not provided :/

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -1,15 +1,54 @@
+  timeWindowDuration: Duration
+): Omit<PacedTrain, 'exceptions'> => ({
+  ...trainSchedule,
+  train_name: pacedTrainId,
+  comfort: 'STANDARD',
```
Why do we need to change the train name and comfort here?

achrafmohye replied:
We enforce `comfort: 'STANDARD'` as it’s required and simplifies imports, i think the choice of the value was disucussed with @maelysLeratRosso as for `train_name` i think it always used `pacedTrainId`

emersion replied:
The comfort is already set elsewhere though? Why repeat it here?

I don't think the train name should contain `pacedTrainId`. `pacedTrainId` is an internal identifier which shouln't be displayed to the user. We already set the train name from the XML file's attributes - why do we need to change it here?

achrafmohye replied:
Where do you see it set elsewhere, i see it set only once at least while the parsing of the xml

emersion replied:
Oh, apologies, I mixed things up with `constraint_distribution`!

And yeah, the `train_name` was the paced train ID in the old code, which is quite weird and confusing… But orthogonal to this PR indeed.

So, why do we need to specify the `constraint_distribution` here? I think stuff should work fine without it?

achrafmohye replied:
The call fails of no `constraint_distribution` is provided

emersion replied:
It seems like `constraint_distribution` has been removed here.

emersion reviewed (commented):
Here's a first round of reviews. I haven't had time to review `reconcilePacedTrainOccurrences()` yet, but will do in the coming days. 

Can the Git history be reworked? Currently some commits introduce some code later removed by subsequent commits.

I'd suggest writing one commit to fix the `buildStepsFromOcp` bug (since it's not really tied to exceptions), and another commit for the rest.

Can we avoid the name "Viriato" in code and commit titles/descriptions? It's an internal SNCF tool.

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -1,15 +1,45 @@
+// These exceptions fields are not handled for now during import
+const fieldsToIgnore = [
+  'train_name',
+  'constraint_distribution',
+  'initial_speed',
+  'labels',
+  'options',
+  'rolling_stock_category',
+  'speed_limit_tag',
```
Do we really need to specify all of these fields? If we always set them to `undefined` during the import, we shouldn't need to blacklist them I think, right?

I think we only need `train_name` here?

Dropping these would avoid missing exceptions for these if we start populating them in the import code.

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -1,15 +1,45 @@
+  'labels',
+  'options',
+  'rolling_stock_category',
+  'speed_limit_tag',
```
Do we really need to specify all of these fields? If we always set them to `undefined` during the import, we shouldn't need to blacklist them I think, right?

I think we only need `train_name` here?

Dropping these would avoid missing exceptions for these if we start populating them in the import code.

emersion replied:
Seems like this has already been taken into account (I accidentally posted this as a standalone comment a few days ago)!

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +102,178 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+
+  const osrdDefaultOccurrences: {
+    startTime: Date;
+    index: number;
```
Do we need to store the `index` here? Can we use the second `forEach` argument instead?

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +102,178 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+
+  for (let i = 0; i < numberOfExpectedOccurrences; i += 1) {
+    const expectedDepartureTime = new Date(
+      firstImportedDepartureTime.getTime() + i * intervalMinutes * 60 * 1000
```
Instead of open-coding some date math, can we use `Duration` instead?

```ts
addDurationToDate(firstImportedDepartureTime, intervalDuration)
```

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +91,177 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+      });
+    } else {
+      // Case 2: OSRD default occurrence has a correspondence in imported file
+      const updatedOccurrence = trainScheduleToPacedTrain(
```
It sounds a bit weird to build a paced train from the train schedule just to be able to check if it has differences with the template paced train. 

Could we change the `updatedOccurrence` argument type in `generatePacedTrainException()` to be `TrainSchedule` instead of `PacedTrain`? The paced train specific fields from that variable aren't used in that function.

achrafmohye replied:
Good point! I explored changing `updatedOccurrence `to `TrainSchedule`, but kept the current approach since it serves mainly to centralize the setting of `comfort `and `train_name `fields that need to be standardized for the comparison.

emersion replied:
I'm not sure I understand the motivation here. I would really rather not build a paced train:

- The `train_name` doesn't matter anyways, since we ignore it below with `omit(exception, 'train_name')`.
- `comfort` should be left unset, just like it was before this PR.

achrafmohye replied:
if we remove the comfort this would result like for the `margins ` in an error in the `rolling_stock: missing field 'comfort'` especially as we are using now the `generatePacedTrainException`

achrafmohye replied:
Done

emersion reviewed (commented):
Sorry, once again I didn't have time to do a full review pass…! Nevertheless, here are some comments!

Thanks for bearing with me, I feel like we're getting closer :)

emersion reviewed (commented):
_Duplicate comment, GitHub won't let me delete it._

emersion commented on front/src/modules/timetableItem/components/ManageTimetableItem/helpers/buildPacedTrainException.ts:
```diff
@@ -125,14 +109,13 @@ export function generatePacedTrainException(
   pacedTrainStartTime.setMilliseconds(0);
 
-  if (!isEqual(originalStartTimeToTest, pacedTrainStartTime)) {
+  if (occurrenceIndex === null || !isEqual(originalStartTimeToTest, pacedTrainStartTime)) {
```
Why add an `occurrenceIndex === null` check here?

An added exception (ie. one without an `occurrenceIndex`) is at least an exception for `start_time`. This change breaks this.

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -138,11 +326,10 @@ const parseXML = async (xmlDoc: Document): Promise<TimetableJsonPayload> => {
+      constraint_distribution: 'STANDARD',
       path,
       schedule,
+      margins: { boundaries: [], values: ['0%'] },
```
Continuing the discussion from https://github.com/OpenRailAssociation/osrd/pull/12498#discussion_r2257484553:

I still believe the margins should be unspecified here. What is the error returned by the API?

We didn't specify margins before this PR, so it sounds weird that these are required.

achrafmohye replied:
here is the error `missing field 'margins' at line 1 column`  return by the call `paced_trains`

achrafmohye replied:
Done :)

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +91,175 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+      };
+
+      const cleanException = omit(exception, 'train_name');
+      exceptions.push(cleanException);
```
Shouldn't we check whether the exception is empty? If `baseException` contains only `train_name`, we can skip it?

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +91,175 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+  }
+
+  const matchedImportedScheduleKeys = new Set<string>();
+  const getScheduleKey = (s: TrainSchedule) => `${s.start_time}-${s.train_name}`;
```
`s.train_name` is identical for all trains, so we can omit it, I think?

achrafmohye replied:
Done

emersion commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +91,175 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+
+    availableImportedSchedules.forEach((importedSchedule) => {
+      const importedDepartureTime = new Date(importedSchedule.start_time);
+      const timeDifferenceMs = Math.abs(
+        osrdOccurrence.startTime.getTime() - importedDepartureTime.getTime()
+      );
+      const timeDifferenceMinutes = timeDifferenceMs / (1000 * 60);
```
Math operations can be avoided with `Duration`:

```ts
Duration.subtractDate(osrdOccurrence.startTime, importedDepartureTime).total('minute')
```

achrafmohye replied:
Done

emersion reviewed (commented):
Thanks for the updated PR!

clarani commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +90,169 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+      totalTimeWindow
+    );
+
+    const baseException = generatePacedTrainException(updatedOccurrence, originalPacedTrain, null);
```
Do you really need to build a pacedTrain here ? What about directly pass the importedSchedule ?
```suggestion
    const baseException = generatePacedTrainException(importedSchedule, originalPacedTrain, null);
```

clarani commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +90,169 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+  );
+
+  sortedImportedSchedules.forEach((importedSchedule) => {
+    if (!unmatchedImportedScheduleKeys.has(getScheduleKey(importedSchedule))) {
```
Hum I think you can check whether `matchedImportedScheduleKeys` has `getScheduleKey(importedSchedule)` or not, and drop `unmatchedImportedScheduleKeys`

clarani commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +90,169 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+      }
+    });
+
+    if (bestCandidate) {
```
I think you can directly create the exception here

clarani commented on front/src/modules/timetableItem/components/ImportTimetableItem/helpers/parseXML.ts:
```diff
@@ -72,6 +90,169 @@ export const getMostFrequentInterval = (schedules: TrainSchedule[]): Duration =>
+        disabled: true,
+      });
+    } else {
+      // Case 2: OSRD default occurrence has a correspondence in imported file
```
```suggestion
      // Case 2: OSRD default occurrence has a match in imported file
```

clarani reviewed (commented):
Great work ! Left some comments

Can the Git history be reworked? Currently some commits introduce some code later removed by subsequent commits.

achrafmohye committed:
front: fix arrival date format

Signed-off-by: Achrafmohye <a.mohyeddine@gmail.com>

achrafmohye committed:
front: make standard as the default distribution constraint

Signed-off-by: Achrafmohye <a.mohyeddine@gmail.com>

achrafmohye committed:
front: compare default occurrences to imported trains and add exceptions
- Mark unmatched default occurrences as disabled
- Mark matched ones as modified if they differ from the mission model
- Ensure each train is matched only once

Signed-off-by: Achrafmohye <a.mohyeddine@gmail.com>

clarani reviewed (approved):
LGTM, tested :white_check_mark:

RomainValls reviewed (approved):
Tested, LGTM 😄
